test/test_known_hosts_cached.rb.rb

```diff
@@ -0,0 +1,208 @@
+  def test_search_for_with_hostname_only_without_check_host_ip
+    options = { user_known_hosts_file: path("known_hosts/gitlab"), check_host_ip: false }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('gitlab.com', options)
+    assert_equal(1, keys.count)
+  end
+
+  def test_search_for_with_ip_only_with_check_host_ip
+    options = { user_known_hosts_file: path("known_hosts/gitlab"), check_host_ip: true }
+    # keys = Net::SSH::KnownHosts.search_for('',options)
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('35.231.145.151', options)
+    assert_equal(1, keys.count)
+  end
+
+  def test_search_for_with_ip_only_without_check_host_ip
+    options = { user_known_hosts_file: path("known_hosts/gitlab"), check_host_ip: false }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('35.231.145.151', options)
+    assert_equal(1, keys.count)
+  end
+
+  def test_search_for_with_hostname_matching_pattern
+    options = { user_known_hosts_file: path("known_hosts/misc") }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('subdomain.gitfoo.com', options)
+    assert_equal(1, keys.count)
+  end
+
+  def test_search_for_with_hostname_not_matching_pattern_1
+    options = { user_known_hosts_file: path("known_hosts/misc") }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('gitfoo.com', options)
+    assert_equal(0, keys.count)
+  end
+
+  def test_search_for_with_hostname_not_matching_pattern_2
+    options = { user_known_hosts_file: path("known_hosts/misc") }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('subdomain.gitmisc.com', options)
+    assert_equal(0, keys.count)
+  end
+
+  def test_search_for_with_hostname_not_matching_pattern_3
+    options = { user_known_hosts_file: path("known_hosts/misc") }
+    kh = Net::SSH::KnownHosts::Cached.new(options)
+    keys = kh.search_for('subsubdomain.subdomain.gitfoo.com', options)
+    assert_equal(0, keys.count)
+  end
+
+  def test_search_for_then_add
+    Tempfile.open('github') do |f|
+      f.write(File.read(path("known_hosts/github")))
+      options = { user_known_hosts_file: f.path }
+      kh = Net::SSH::KnownHosts::Cached.new(build_opts(f.path))
+      keys = kh.search_for('github2.com', options)
+      assert_equal(0, keys.count)
+
+      keys.add_host_key(rsa_key)
+
+      assert_equal([rsa_key.to_blob], keys.map(&:to_blob))
+      keys = kh.search_for('github2.com', options)
+      assert_equal([rsa_key.to_blob], keys.map(&:to_blob))
+    end
+  end
+
+  def test_host_entry_added_out_of_process
+    Tempfile.open('github') do |f|
+      options = { user_known_hosts_file: f.path }
+      kh = Net::SSH::KnownHosts::Cached.new(build_opts(f.path))
+      keys = kh.search_for('github.com', options)
+      assert_equal(0, keys.count)
+      f.write(File.read(path("known_hosts/github")))
+      f.close
+      # Ensure mtime is different
```

Comment:
This is where I found issues with mtimes being so close that the file would not be re-parsed.